apps/emqx/src/emqx_tls_certfile_gc.erl

```diff
@@ -0,0 +1,335 @@
+    {ok, _Events :: [event()]}.
+run() ->
+    gen_server:call(?MODULE, collect).
+
+-spec force() ->
+    {ok, _Events :: [event()]}.
+force() ->
+    % NOTE
+    % Simulate a complete GC cycle by running it twice. Mostly useful in tests.
+    {ok, Events1} = run(),
+    {ok, Events2} = run(),
+    {ok, Events1 ++ Events2}.
+
+%%--------------------------------------------------------------------
+%% Supervisor callbacks
+%%--------------------------------------------------------------------
+
+-spec init(_) ->
+    {ok, st()}.
+init(Interval) ->
+    {ok, start_timer(#{gc_interval => Interval})}.
+
+-spec handle_call(collect | _Call, gen_server:from(), st()) ->
+    {reply, {ok, [event()]}, st()} | {noreply, st()}.
+handle_call(collect, From, St) ->
+    {ok, Events, StNext} = ?tp_span(
+        tls_certfile_gc_manual,
+        #{caller => From},
+        collect(St, #{evhandler => {fun emqx_utils:cons/2, []}})
+    ),
+    {reply, {ok, Events}, restart_timer(StNext)};
+handle_call(Call, From, St) ->
+    ?SLOG(error, #{msg => "unexpected_call", call => Call, from => From}),
+    {noreply, St}.
+
+-spec handle_cast(_Cast, st()) ->
+    {noreply, st()}.
+handle_cast(Cast, State) ->
+    ?SLOG(error, #{msg => "unexpected_cast", cast => Cast}),
+    {noreply, State}.
+
+-spec handle_info({timeout, reference(), collect}, st()) ->
+    {noreply, st()}.
+handle_info({timeout, TRef, collect}, St = #{next_gc_timer := TRef}) ->
+    {ok, _, StNext} = ?tp_span(
+        tls_certfile_gc_periodic,
+        #{},
+        collect(St, #{evhandler => {fun log_event/2, []}})
+    ),
+    {noreply, start_timer(StNext)}.
+
+start_timer(St = #{gc_interval := Interval}) ->
+    TRef = erlang:start_timer(Interval, self(), collect),
+    St#{next_gc_timer => TRef}.
+
+restart_timer(St = #{next_gc_timer := TRef}) ->
+    ok = emqx_utils:cancel_timer(TRef),
+    start_timer(St).
+
+log_event({collect, Filename, ok}, _) ->
+    ?tp(debug, "tls_certfile_gc_collected", #{filename => Filename});
+log_event({collect, Filename, {error, Reason}}, _) ->
+    ?tp(warning, "tls_certfile_gc_collect_error", #{filename => Filename, reason => Reason}).
+
+%%--------------------------------------------------------------------
+%% Internal functions
+%% -------------------------------------------------------------------
+
+collect(St, Opts) ->
+    RootDir = emqx_utils_fs:canonicalize(emqx:mutable_certs_dir()),
+    Orphans = orphans(RootDir),
+    OrphansLast = maps:get(orphans, St, #{}),
+    Convicts = convicts(Orphans, OrphansLast),
+    Result = collect_files(Convicts, RootDir, Opts),
+    {ok, Result, St#{orphans => maps:without(Convicts, Orphans)}}.
+
+orphans() ->
+    Dir = emqx_utils_fs:canonicalize(emqx:mutable_certs_dir()),
+    orphans(Dir).
+
+orphans(Dir) ->
+    Certfiles = find_managed_files(fun is_managed_file/2, Dir),
+    lists:foldl(
+        fun(Root, Acc) ->
+            References = find_references(Root),
+            maps:without(References, Acc)
```

Comment:
if the certs directory is once symlinked and expanded to an abs path,
e.g. `foo.bar.ssl_options.cacertfile=/data1/certs/cacertfile-0000.pem`
then  `node.data_dir` changes to another link, e.g. `/data2/` (node restart), this will cause all the cert files to be deleted after 2 collection-ticks?

Reply:
`data1/certs/bridge1/cacertfile-0000.pem`
`data1/certs/bridge2/cacertfile-0000.pem`

Reply:
solution: use random suffix instead of cert hash hex string for file name suffix.